### DISCUSSION TITLE
Assert that exception message contains one of multiple strings

### DISCUSSION THREAD

**Discussion by sebthom** [assertj/assertj#2376] [Q&A] [answered]
I have the case where I want assert that the thrown exception contains one of several possible strings in the error message.

What is the best way to achieve something like this:

```java
assertThatThrownBy(() -> {
  /* code to test */
})
.isInstanceOf(CertPathValidatorException.class)
.hasMessageContaining(
   either( // this is pseudo code
     "could not be determined", 
     "could not determine revocation status", 
     "revocation status check failed"
   )
)
```

Another case would be
```java
assertThatThrownBy(() -> {
  /* code to test */
})
.isInstanceOf(CertPathValidatorException.class)
.hasMessageContaining(
   // pseudo code
   System.getProperty("java.vm.vendor").contains("IBM") 
     ? "has been revoked for an unspecified reason"
     : either(
         "could not determine revocation status", 
         "revocation status check failed"
       )
)
```

**Comment by scordio** [comment, marked as answer]
[`hasMessageMatching`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractThrowableAssert.html#hasMessageMatching(java.lang.String)) might be an alternative for some cases. Not the best readability, though.

Another option that can be used is [`extracting(Function, InstanceOfAssertFactory)`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractObjectAssert.html#extracting(java.util.function.Function,org.assertj.core.api.InstanceOfAssertFactory)):

```java
assertThatThrownBy(() -> {
  /* code to test */
})
  .isInstanceOf(CertPathValidatorException.class)
  .extracting(Throwable::getMessage, as(InstanceOfAssertFactories.STRING))
  .containsAnyOf(
    "could not be determined",
    "could not determine revocation status",
    "revocation status check failed"
  );
```

We could add a navigational method to get the message as a string assertion, that would shorten a bit the `extracting` line. It would allow writing something like:

```java
assertThatThrownBy(() -> {
  /* code to test */
})
  .isInstanceOf(CertPathValidatorException.class)
  .message() // returns StringAssert
  .containsAnyOf(
    "could not be determined", 
    "could not determine revocation status", 
    "revocation status check failed"
  );
```

Readability-wise, I'm not sure what is the best way to handle the conditional string under `System.getProperty("java.vm.vendor").contains("IBM")`. Maybe you could declare the `containsAnyOf` input outside the assertion as a `CharSequence` array, which would have one element if the condition is true or two elements if the condition is false:

```java
CharSequence[] messages = System.getProperty("java.vm.vendor").contains("IBM")
  ? new CharSequence[] { "has been revoked for an unspecified reason" }
  : new CharSequence[] { "could not determine revocation status", "revocation status check failed" };

assertThatThrownBy(() -> {
  /* code to test */
})
  .isInstanceOf(CertPathValidatorException.class)
  .extracting(Throwable::getMessage, as(InstanceOfAssertFactories.STRING))
  .containsAnyOf(messages);
```

**Reply by sebthom** [reply]
Thanks your quick response! Very valuable information!

I am using your last suggestion now.

It would be nice if the `message()` shortcut for `.extracting(Throwable::getMessage, as(InstanceOfAssertFactories.STRING))` you mentioned could be added to `AbstractThrowableAssert`.

**Reply by scordio** [reply]
Thanks for raising #2378!
